File: rust/src/tools/agent.rs
//! Agent CLI tool configuration (@link-assistant/agent).
//!
//! Based on hive-mind's agent.lib.mjs implementation.
//! Agent is a fork of OpenCode that ships a native, enforceable permission
//! system (agent v0.24.0, PR #272) exposed through `--permission-mode`
//! (auto | plan | readonly | ask) and an OpenCode-compatible `--permission`
//! JSON policy.

use crate::streaming::parse_ndjson;
use crate::tools::shell::{build_command_head, escape_arg, escape_single_quotes};
use serde_json::Value;
use std::collections::HashMap;

/// Get the Agent model map
/// Maps aliases to full model IDs (uses OpenCode's provider/model format)
pub fn get_model_map() -> HashMap<&'static str, &'static str> {
    let mut map = HashMap::new();
    // OpenCode Zen free models (current)
    map.insert("grok", "opencode/grok-code");
    map.insert("grok-code", "opencode/grok-code");
    map.insert("grok-code-fast-1", "opencode/grok-code");
    map.insert("big-pickle", "opencode/big-pickle");
    map.insert("gpt-5-nano", "opencode/gpt-5-nano");
    map.insert("minimax-m2.5-free", "opencode/minimax-m2.5-free");
    // Default: NVIDIA hybrid Mamba-Transformer (hive-mind issue #1563, agent PR #243)
    map.insert("nemotron-3-super-free", "opencode/nemotron-3-super-free");
    // Kilo Gateway free models
    map.insert("glm-5-free", "kilo/glm-5-free");
    map.insert("glm-4.5-air-free", "kilo/glm-4.5-air-free");
    map.insert("deepseek-r1-free", "kilo/deepseek-r1-free");
    map.insert("giga-potato-free", "kilo/giga-potato-free");
    map.insert("trinity-large-preview", "kilo/trinity-large-preview");
    // Deprecated free models (kept for backward compatibility)
    map.insert("qwen3.6-plus-free", "opencode/qwen3.6-plus-free"); // Deprecated: free promotion ended April 2026
    map.insert("kimi-k2.5-free", "opencode/kimi-k2.5-free");
    map.insert("glm-4.7-free", "opencode/glm-4.7-free");
    map.insert("minimax-m2.1-free", "opencode/minimax-m2.1-free");
    // Premium models
    map.insert("sonnet", "anthropic/claude-3-5-sonnet");
    map.insert("haiku", "anthropic/claude-3-5-haiku");
    map.insert("opus", "anthropic/claude-3-opus");
    map.insert("gemini-3-pro", "google/gemini-3-pro");
    map
}

/// Map model alias to full model ID
///
/// # Arguments
/// * `model` - Model alias or full ID
///
/// # Returns
/// Full model ID
pub fn map_model_to_id(model: &str) -> String {
    let model_map = get_model_map();
    model_map
        .get(model)
        .map(|s| s.to_string())
        .unwrap_or_else(|| model.to_string())
}

/// Agent command build options
#[derive(Debug, Clone, Default)]
pub struct AgentBuildOptions {
    pub prompt: Option<String>,
    pub prompt_file: Option<String>,
    pub system_prompt: Option<String>,
    pub model: Option<String>,
    pub compact_json: bool,
    pub use_existing_claude_oauth: bool,
    /// Enforce hard read-only mode (`--permission-mode readonly`)
    pub read_only: bool,
    /// Enforce planning mode (`--permission-mode plan`)
    pub plan_only: bool,
    /// Explicit agent permission mode (auto | plan | readonly | ask)
    pub permission_mode: Option<String>,
    /// OpenCode-compatible `--permission` JSON policy
    pub permission: Option<String>,
    pub executable: Option<String>,
    pub extra_env: Vec<(String, String)>,
    pub extra_args: Vec<String>,
}

/// Build command line arguments for Agent
///
/// # Arguments
/// * `options` - Build options
///
/// # Returns
/// Vector of CLI arguments
pub fn build_args(options: &AgentBuildOptions) -> Vec<String> {
    let mut args = Vec::new();

    // Native, enforceable permission system (agent v0.24.0, PR #272).
    // --plan-only maps to `plan`, --read-only maps to the harder `readonly`,
    // matching agent's own distinction between the two modes.
    let resolved_permission_mode = options.permission_mode.clone().or_else(|| {
        if options.plan_only {
            Some("plan".to_string())
        } else if options.read_only {
            Some("readonly".to_string())
        } else {
            None
        }
    });
    if let Some(mode) = resolved_permission_mode {
        args.push("--permission-mode".to_string());
        args.push(mode);
    }

    if let Some(ref permission) = options.permission {
        args.push("--permission".to_string());
        args.push(permission.clone());
    }

    if let Some(ref model) = options.model {
        let mapped_model = map_model_to_id(model);
        args.push("--model".to_string());
        args.push(mapped_model);
    }

    if options.compact_json {
        args.push("--compact-json".to_string());
    }

    if options.use_existing_claude_oauth {
        args.push("--use-existing-claude-oauth".to_string());
    }

    args.extend(options.extra_args.clone());

    args
}

/// Build complete command string for Agent
/// Agent uses stdin for prompt input (NDJSON streaming supported)
///
/// # Arguments
/// * `options` - Build options
///
/// # Returns
/// Complete command string
pub fn build_command(options: &AgentBuildOptions) -> String {
    let args = build_args(options);
    let args_str: Vec<String> = args.iter().map(|a| escape_arg(a)).collect();

    // Agent expects prompt via stdin, combine system and user prompts
    let combined_prompt = match (&options.system_prompt, &options.prompt) {
        (Some(sys), Some(prompt)) => format!("{}\n\n{}", sys, prompt),
        (Some(sys), None) => sys.clone(),
        (None, Some(prompt)) => prompt.clone(),
        (None, None) => String::new(),
    };

    // Build command with stdin piping
    let input_command = options.prompt_file.as_ref().map_or_else(
        || format!("printf '%s' '{}'", escape_single_quotes(&combined_prompt)),
        |prompt_file| format!("cat {}", escape_arg(prompt_file)),
    );
    let executable = options.executable.as_deref().unwrap_or("agent");
    format!(
        "{} | {} {}",
        input_command,
        build_command_head(executable, &options.extra_env, &[]),
        args_str.join(" ")
    )
    .trim()
    .to_string()
}

/// Parse JSON messages from Agent output
/// Agent outputs NDJSON format with specific event types
///
/// # Arguments
/// * `output` - Raw output string
///
/// # Returns
/// Vector of parsed JSON messages
pub fn parse_output(output: &str) -> Vec<Value> {
    parse_ndjson(output)
}

/// Extract session ID from Agent output
///
/// # Arguments
/// * `output` - Raw output string
///
/// # Returns
/// Session ID or None
pub fn extract_session_id(output: &str) -> Option<String> {
    let messages = parse_output(output);

    for msg in messages {
        if let Some(session_id) = msg.get("session_id").and_then(|v| v.as_str()) {
            return Some(session_id.to_string());
        }
    }

    None
}

/// Usage statistics for Agent
#[derive(Debug, Clone, Default)]
pub struct AgentUsage {
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub reasoning_tokens: u64,
    pub cache_read_tokens: u64,
    pub cache_write_tokens: u64,
    pub total_cost: f64,
    pub step_count: u64,
}

/// Parse token usage from Agent output
/// Agent outputs step_finish events with token data
///
/// # Arguments
/// * `output` - Raw output string
///
/// # Returns
/// Token usage statistics
pub fn extract_usage(output: &str) -> AgentUsage {
    let messages = parse_output(output);
    let mut usage = AgentUsage::default();

    for msg in messages {
        // Look for step_finish events which contain token usage
        if msg.get("type").and_then(|t| t.as_str()) == Some("step_finish") {
            if let Some(part) = msg.get("part") {
                usage.step_count += 1;

                if let Some(tokens) = part.get("tokens") {
                    if let Some(input) = tokens.get("input").and_then(|v| v.as_u64()) {
                        usage.input_tokens += input;
                    }
                    if let Some(output) = tokens.get("output").and_then(|v| v.as_u64()) {
                        usage.output_tokens += output;
                    }
                    if let Some(reasoning) = tokens.get("reasoning").and_then(|v| v.as_u64()) {
                        usage.reasoning_tokens += reasoning;
                    }

                    // Handle cache tokens
                    if let Some(cache) = tokens.get("cache") {
                        if let Some(read) = cache.get("read").and_then(|v| v.as_u64()) {
                            usage.cache_read_tokens += read;
                        }
                        if let Some(write) = cache.get("write").and_then(|v| v.as_u64()) {
                            usage.cache_write_tokens += write;
                        }
                    }
                }

                // Add cost from step_finish
                if let Some(cost) = part.get("cost").and_then(|v| v.as_f64()) {
                    usage.total_cost += cost;
                }
            }
        }
    }

    usage
}

/// Error detection result
#[derive(Debug, Clone, Default)]
pub struct ErrorResult {
    pub has_error: bool,
    pub error_type: Option<String>,
    pub message: Option<String>,
}

/// Detect errors in Agent output
///
/// # Arguments
/// * `output` - Raw output string
///
/// # Returns
/// Error detection result
pub fn detect_errors(output: &str) -> ErrorResult {
    let messages = parse_output(output);

    for msg in messages {
        // Check for explicit error message types from agent
        let msg_type = msg.get("type").and_then(|t| t.as_str());
        if msg_type == Some("error") || msg_type == Some("step_error") {
            return ErrorResult {
                has_error: true,
                error_type: msg_type.map(|s| s.to_string()),
                message: msg
                    .get("message")
                    .and_then(|m| m.as_str())
                    .map(|s| s.to_string())
                    .or_else(|| Some("Unknown error".to_string())),
            };
        }
    }

    ErrorResult::default()
}

/// Agent tool configuration
#[derive(Debug, Clone)]
pub struct AgentTool {
    pub name: &'static str,
    pub display_name: &'static str,
    pub executable: &'static str,
    pub supports_json_output: bool,
    pub supports_json_input: bool,
    pub supports_system_prompt: bool,
    pub supports_resume: bool,
    pub supports_read_only: bool,
    pub default_model: &'static str,
}

impl Default for AgentTool {
    fn default() -> Self {
        Self {
            name: "agent",
            display_name: "@link-assistant/agent",
            executable: "agent",
            supports_json_output: true,
            supports_json_input: true, // Agent supports full JSON streaming input
            supports_system_prompt: false, // System prompt is combined with user prompt
            supports_resume: false,    // Agent doesn't have explicit resume like Claude
            supports_read_only: true, // Native --permission-mode readonly/plan (agent v0.24.0, PR #272)
            default_model: "nemotron-3-super-free", // hive-mind issue #1563, agent PR #243
        }
    }
}

// Tests are in rust/tests/agent_tests.rs

File: js/src/tools/agent.mjs
/**
 * Agent CLI tool configuration (@link-assistant/agent)
 * Based on hive-mind's agent.lib.mjs implementation
 * Agent is a fork of OpenCode that ships a native, enforceable permission
 * system (agent v0.24.0, PR #272) exposed through `--permission-mode`
 * (auto | plan | readonly | ask) and an OpenCode-compatible `--permission`
 * JSON policy.
 */

import { buildCommandHead, escapeArg, normalizeExtraArgs } from './shell.mjs';

/**
 * Available Agent model configurations
 * Maps aliases to full model IDs (uses OpenCode's provider/model format)
 */
export const modelMap = {
  // OpenCode Zen free models (current)
  grok: 'opencode/grok-code',
  'grok-code': 'opencode/grok-code',
  'grok-code-fast-1': 'opencode/grok-code',
  'big-pickle': 'opencode/big-pickle',
  'gpt-5-nano': 'opencode/gpt-5-nano',
  'minimax-m2.5-free': 'opencode/minimax-m2.5-free',
  // Default: NVIDIA hybrid Mamba-Transformer (hive-mind issue #1563, agent PR #243)
  'nemotron-3-super-free': 'opencode/nemotron-3-super-free',
  // Kilo Gateway free models
  'glm-5-free': 'kilo/glm-5-free',
  'glm-4.5-air-free': 'kilo/glm-4.5-air-free',
  'deepseek-r1-free': 'kilo/deepseek-r1-free',
  'giga-potato-free': 'kilo/giga-potato-free',
  'trinity-large-preview': 'kilo/trinity-large-preview',
  // Full names with kilo/ prefix
  'kilo/glm-5-free': 'kilo/glm-5-free',
  'kilo/glm-4.5-air-free': 'kilo/glm-4.5-air-free',
  'kilo/minimax-m2.5-free': 'kilo/minimax-m2.5-free',
  'kilo/deepseek-r1-free': 'kilo/deepseek-r1-free',
  'kilo/giga-potato-free': 'kilo/giga-potato-free',
  'kilo/trinity-large-preview': 'kilo/trinity-large-preview',
  // Deprecated free models (kept for backward compatibility)
  'qwen3.6-plus-free': 'opencode/qwen3.6-plus-free', // Deprecated: free promotion ended April 2026
  'kimi-k2.5-free': 'opencode/kimi-k2.5-free',
  'glm-4.7-free': 'opencode/glm-4.7-free',
  'minimax-m2.1-free': 'opencode/minimax-m2.1-free',
  'kilo/glm-4.7-free': 'kilo/glm-4.7-free',
  'kilo/kimi-k2.5-free': 'kilo/kimi-k2.5-free',
  'kilo/minimax-m2.1-free': 'kilo/minimax-m2.1-free',
  // Premium models
  sonnet: 'anthropic/claude-3-5-sonnet',
  haiku: 'anthropic/claude-3-5-haiku',
  opus: 'anthropic/claude-3-opus',
  'gemini-3-pro': 'google/gemini-3-pro',
};

/**
 * Map model alias to full model ID
 * @param {Object} options - Options
 * @param {string} options.model - Model alias or full ID
 * @returns {string} Full model ID
 */
export function mapModelToId(options) {
  const { model } = options;
  return modelMap[model] || model;
}

/**
 * Build command line arguments for Agent
 * @param {Object} options - Options
 * @param {string} [options.prompt] - User prompt
 * @param {string} [options.systemPrompt] - System prompt (combined with user prompt)
 * @param {string} [options.model] - Model to use
 * @param {boolean} [options.compactJson] - Use compact JSON output
 * @param {boolean} [options.useExistingClaudeOAuth] - Use existing Claude OAuth credentials
 * @param {boolean} [options.readOnly] - Enforce hard read-only mode (`--permission-mode readonly`)
 * @param {boolean} [options.planOnly] - Enforce planning mode (`--permission-mode plan`)
 * @param {boolean} [options.approveEach] - Approve each mutating command (`--permission-mode ask`)
 * @param {string} [options.permissionMode] - Explicit agent permission mode (auto | plan | readonly | ask)
 * @param {string} [options.permission] - OpenCode-compatible `--permission` JSON policy
 * @param {string[]} [options.extraArgs] - Extra raw CLI args appended after typed args
 * @returns {string[]} Array of CLI arguments
 */
export function buildArgs(options) {
  const {
    model,
    compactJson = false,
    useExistingClaudeOAuth = false,
    readOnly = false,
    planOnly = false,
    approveEach = false,
    permissionMode,
    permission,
    extraArgs = [],
  } = options;

  const args = [];

  // Native, enforceable permission system (agent v0.24.0, PR #272).
  // --plan-only maps to `plan`, --read-only maps to the harder `readonly`,
  // --approve-each maps to `ask` (per-command approval relayed over JSON),
  // matching agent's own distinction between the modes. An explicit
  // permissionMode always wins.
  const resolvedPermissionMode =
    permissionMode ||
    (approveEach
      ? 'ask'
      : planOnly
        ? 'plan'
        : readOnly
          ? 'readonly'
          : undefined);
  if (resolvedPermissionMode) {
    args.push('--permission-mode', resolvedPermissionMode);
  }

  // Ask mode emits requests mid-turn and blocks until answered, so it requires
  // a streaming input mode (single-shot --prompt would deadlock).
  if (resolvedPermissionMode === 'ask') {
    args.push('--input-format', 'stream-json');
  }

  if (permission) {
    args.push('--permission', permission);
  }

  if (model) {
    const mappedModel = mapModelToId({ model });
    args.push('--model', mappedModel);
  }

  if (compactJson) {
    args.push('--compact-json');
  }

  if (useExistingClaudeOAuth) {
    args.push('--use-existing-claude-oauth');
  }

  args.push(...normalizeExtraArgs(extraArgs));

  return args;
}

/**
 * Build complete command string for Agent
 * Agent uses stdin for prompt input (NDJSON streaming supported)
 * @param {Object} options - Options
 * @param {string} options.workingDirectory - Working directory
 * @param {string} [options.prompt] - User prompt
 * @param {string} [options.promptFile] - File containing combined prompt input
 * @param {string} [options.systemPrompt] - System prompt
 * @param {string} [options.model] - Model to use
 * @param {boolean} [options.compactJson] - Use compact JSON output
 * @param {boolean} [options.useExistingClaudeOAuth] - Use existing Claude OAuth
 * @param {boolean} [options.readOnly] - Enforce hard read-only mode (`--permission-mode readonly`)
 * @param {boolean} [options.planOnly] - Enforce planning mode (`--permission-mode plan`)
 * @param {boolean} [options.approveEach] - Approve each mutating command (`--permission-mode ask`)
 * @param {string} [options.permissionMode] - Explicit agent permission mode (auto | plan | readonly | ask)
 * @param {string} [options.permission] - OpenCode-compatible `--permission` JSON policy
 * @param {string} [options.executable='agent'] - Executable path/name
 * @param {Object|Array} [options.extraEnv] - Environment variables for the tool
 * @param {string[]} [options.extraArgs] - Extra raw CLI args appended after typed args
 * @returns {string} Complete command string
 */
export function buildCommand(options) {
  const {
    prompt,
    promptFile,
    systemPrompt,
    executable = 'agent',
    extraEnv,
    streamInput = false,
    ...argOptions
  } = options;
  const args = buildArgs(argOptions);

  const commandHead = `${buildCommandHead({
    executable,
    extraEnv,
  })} ${args.map(escapeArg).join(' ')}`.trim();

  // In stream-input mode the caller owns the child's stdin and writes the
  // prompt and permission responses as NDJSON frames (per-command approval
  // relay), so no prompt is piped here.
  if (streamInput) {
    return commandHead;
  }

  // Agent expects prompt via stdin, combine system and user prompts
  const combinedPrompt = systemPrompt
    ? `${systemPrompt}\n\n${prompt || ''}`
    : prompt || '';

  // Build command with stdin piping
  const inputCommand = promptFile
    ? `cat ${escapeArg(promptFile)}`
    : `printf '%s' '${combinedPrompt.replace(/'/g, "'\\''")}'`;
  return `${inputCommand} | ${commandHead}`.trim();
}

/**
 * Parse JSON messages from Agent output
 * Agent outputs NDJSON format with specific event types
 * @param {Object} options - Options
 * @param {string} options.output - Raw output string
 * @returns {Object[]} Array of parsed JSON messages
 */
export function parseOutput(options) {
  const { output } = options;
  const messages = [];
  const lines = output.split('\n');

  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed || !trimmed.startsWith('{')) {
      continue;
    }

    try {
      const parsed = JSON.parse(trimmed);
      messages.push(parsed);
    } catch {
      // Skip lines that aren't valid JSON
    }
  }

  return messages;
}

/**
 * Extract session ID from Agent output
 * @param {Object} options - Options
 * @param {string} options.output - Raw output string
 * @returns {string|null} Session ID or null
 */
export function extractSessionId(options) {
  const { output } = options;
  const messages = parseOutput({ output });

  for (const msg of messages) {
    if (msg.session_id) {
      return msg.session_id;
    }
  }

  return null;
}

/**
 * Parse token usage from Agent output
 * Agent outputs step_finish events with token data
 * @param {Object} options - Options
 * @param {string} options.output - Raw output string
 * @returns {Object} Token usage statistics
 */
export function extractUsage(options) {
  const { output } = options;
  const messages = parseOutput({ output });

  const usage = {
    inputTokens: 0,
    outputTokens: 0,
    reasoningTokens: 0,
    cacheReadTokens: 0,
    cacheWriteTokens: 0,
    totalCost: 0,
    stepCount: 0,
  };

  for (const msg of messages) {
    // Look for step_finish events which contain token usage
    if (msg.type === 'step_finish' && msg.part?.tokens) {
      const tokens = msg.part.tokens;
      usage.stepCount++;

      // Add token counts
      if (tokens.input) {
        usage.inputTokens += tokens.input;
      }
      if (tokens.output) {
        usage.outputTokens += tokens.output;
      }
      if (tokens.reasoning) {
        usage.reasoningTokens += tokens.reasoning;
      }

      // Handle cache tokens
      if (tokens.cache) {
        if (tokens.cache.read) {
          usage.cacheReadTokens += tokens.cache.read;
        }
        if (tokens.cache.write) {
          usage.cacheWriteTokens += tokens.cache.write;
        }
      }

      // Add cost from step_finish
      if (msg.part.cost !== undefined) {
        usage.totalCost += msg.part.cost;
      }
    }
  }

  return usage;
}

/**
 * Detect errors in Agent output
 * @param {Object} options - Options
 * @param {string} options.output - Raw output string
 * @returns {Object} Error detection result
 */
export function detectErrors(options) {
  const { output } = options;
  const messages = parseOutput({ output });

  for (const msg of messages) {
    // Check for explicit error message types from agent
    if (msg.type === 'error' || msg.type === 'step_error') {
      return {
        hasError: true,
        errorType: msg.type,
        message: msg.message || 'Unknown error',
      };
    }
  }

  return { hasError: false };
}

/**
 * Agent tool configuration
 */
export const agentTool = {
  name: 'agent',
  displayName: '@link-assistant/agent',
  executable: 'agent',
  supportsJsonOutput: true,
  supportsJsonInput: true, // Agent supports full JSON streaming input
  supportsSystemPrompt: false, // System prompt is combined with user prompt
  supportsResume: false, // Agent doesn't have explicit resume like Claude
  supportsReadOnly: true, // Native --permission-mode readonly/plan (agent v0.24.0, PR #272)
  supportsAsk: true, // Native --permission-mode ask with JSON permission relay
  defaultModel: 'nemotron-3-super-free', // hive-mind issue #1563, agent PR #243
  modelMap,
  mapModelToId,
  buildArgs,
  buildCommand,
  parseOutput,
  extractSessionId,
  extractUsage,
  detectErrors,
};
